src/twisted/conch/ssh/connection.py

```diff
@@ -65,8 +66,16 @@ def serviceStopped(self):
         """
         Called when the connection is stopped.
         """
-        for channel in list(self.channels.values()):
+        # Close any fully open channels
+        for channel in list(self.channelsToRemoteChannel.keys()):
             self.channelClosed(channel)
+        # Indicate failure to any channels that were in the process of
+        # opening but not yet open.
+        while self.channels:
```

Comment:
Good point, and synchronously reopening a channel isn't an unlikely response, either.

I habitually use this pattern because it correctly handles *most* cases of synchronous modifications to the container, both additions and removals. In this case I think the only downside is the possibility of getting in a loop with synchronous re-opens, but IMHO an attempt to open a channel while the connection is closing should simply raise an exception (analogously to [bug 9572](https://twistedmatrix.com/trac/ticket/9572)). I'll look into that and either update this PR or open a separate bug for it.

Reply:
I filed a bug here: https://twistedmatrix.com/trac/ticket/9626 ; I think you're right that it's a possibility. @markrwilliams if you have time to weigh in on that ticket and 9572 I'd appreciate any thoughts you have.